crates/bevy_render/macros/src/as_bind_group.rs

```diff
@@ -0,0 +1,389 @@
+    OccupiedMergableUniform {
+        uniform_fields: Vec<&'a Field>,
+    },
+}
+
+pub fn derive_as_bind_group(input: TokenStream) -> TokenStream {
+    let ast = parse_macro_input!(input as DeriveInput);
+
+    let manifest = BevyManifest::default();
+    let render_path = manifest.get_path("bevy_render");
+    let asset_path = manifest.get_path("bevy_asset");
+
+    let mut binding_states: Vec<BindingState> = Vec::new();
+    let mut binding_impls = Vec::new();
+    let mut bind_group_entries = Vec::new();
+    let mut binding_layouts = Vec::new();
+    let mut attr_prepared_data_ident = None;
+
+    // Read struct-level attributes
+    for attr in &ast.attrs {
+        if let Some(attr_ident) = attr.path.get_ident() {
+            if attr_ident == BIND_GROUP_DATA_ATTRIBUTE_NAME {
+                if let Ok(prepared_data_ident) =
+                    attr.parse_args_with(|input: ParseStream| input.parse::<Ident>())
+                {
+                    attr_prepared_data_ident = Some(prepared_data_ident);
+                }
+            } else if attr_ident == UNIFORM_ATTRIBUTE_NAME {
+                let (binding_index, converted_shader_type) = attr
+                    .parse_args_with(|input: ParseStream| {
+                        let binding_index = input
+                            .parse::<LitInt>()
+                            .and_then(|i| i.base10_parse::<u32>())?;
+                        input.parse::<Comma>()?;
+                        let converted_shader_type = input.parse::<Ident>()?;
+                        Ok((binding_index, converted_shader_type))
+                    })
+                    .unwrap_or_else(|_| {
+                        panic!("struct-level uniform bindings must be in the format: uniform(BINDING_INDEX, ConvertedShaderType)");
+                    });
+
+                binding_impls.push(quote! {{
+                    use #render_path::render_resource::AsBindGroupShaderType;
+                    let mut buffer = #render_path::render_resource::encase::UniformBuffer::new(Vec::new());
+                    let converted: #converted_shader_type = self.as_bind_group_shader_type(images);
+                    buffer.write(&converted).unwrap();
+                    #render_path::render_resource::OwnedBindingResource::Buffer(render_device.create_buffer_with_data(
+                        &#render_path::render_resource::BufferInitDescriptor {
+                            label: None,
+                            usage: #render_path::render_resource::BufferUsages::COPY_DST | #render_path::render_resource::BufferUsages::UNIFORM,
+                            contents: buffer.as_ref(),
+                        },
+                    ))
+                }});
+
+                binding_layouts.push(quote!{
+                    #render_path::render_resource::BindGroupLayoutEntry {
+                        binding: #binding_index,
+                        visibility: #render_path::render_resource::ShaderStages::all(),
```

Comment:
I'm curious about that. Are there performance implications of making them visible to both stages? I'm still pretty new to GPU programming so it might just be me not knowing something. I just assumed this option existed so you could limit how much data is sent to each stage, but at this point I'm not even sure if it makes sense to think that data is sent to a stage.

Reply:
Hard to say! Generally it seems like it shouldn't matter much:
https://github.com/KhronosGroup/Vulkan-Docs/issues/1790

Switching between "all" and "fragment" on many_cubes seems to be "within the noise". Probably worth supporting configuring this, but also probably not the biggest priority.